### PR TITLE
More funcotator fixes

### PR DESCRIPTION
**Pull request by LeeTL1220**
These fixes are addressing issues for annotating germline VCFs.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java [hunk 1]
```diff
@@ -329,14 +335,7 @@ public void onTraversalStart() {
...
         // Sort our data source factories to ensure they're always in the same order:
-        dataSourceFactories.sort( Comparator.comparing(DataSourceFuncotationFactory::getName) );
-
-        // Identify and store any Gencode data sources:
-        for ( final DataSourceFuncotationFactory factory : dataSourceFactories ) {
-            if ( factory.getType().equals(FuncotatorArgumentDefinitions.DataSourceType.GENCODE) ) {
-                gencodeFuncotationFactories.add( (GencodeFuncotationFactory) factory );
-            }
-        }
+        dataSourceFactories.sort(DataSourceUtils::datasourceComparator);
```
Comment (jonn-smith) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java]: Can you add to this comment that gencode factories will always be first as well?

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java [hunk 2]
```diff
@@ -471,8 +470,13 @@ private void enqueueAndHandleVariant(final VariantContext variant, final Referen
             for ( final FeatureInput<? extends Feature> featureInput : manualLocatableFeatureInputs ) {
                 @SuppressWarnings("unchecked")
                 final List<Feature> featureList = (List<Feature>)featureContext.getValues(featureInput, hg19Interval);
+
+                // TODO: This is a little sloppy, since it checks every datasource twice.  Once for hg19 contig names and once for b37 contig names.
+                featureList.addAll(featureContext.getValues(featureInput));
```
Comment (jonn-smith) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java]: Can we not detect the reference and only add in the features with the correct names?

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java [hunk 3]
```diff
@@ -482,34 +486,25 @@ private void enqueueAndHandleVariant(final VariantContext variant, final Referen
...
-            );
-        }
-
-        // Annotate with the rest of the data sources:
-        for ( final DataSourceFuncotationFactory funcotationFactory : dataSourceFactories ) {
-
-            // Make sure we don't double up on the Gencodes:
-            if ( funcotationFactory.getType().equals(FuncotatorArgumentDefinitions.DataSourceType.GENCODE) ) {
-                continue;
+        // Perform the actual annotation.  Note that we leverage the ordering of datasources here.
```
Comment (jonn-smith) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java]: Please add another comment here explicitly calling out that the comparator guarantees that gencode data sources are all first in the list.

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/vcfOutput/VcfOutputRenderer.java [hunk 4]
```diff
@@ -180,13 +182,21 @@ private VCFHeader createVCFHeader() {
...
-        return dataSourceFactories.stream()
+        return dataSourceFactories.stream().sorted(DataSourceUtils::datasourceComparator)
                         .map(DataSourceFuncotationFactory::getSupportedFuncotationFields)
                         .flatMap(LinkedHashSet::stream)
                         .map(Object::toString).collect(Collectors.joining(HEADER_LISTED_FIELD_DELIMITER));
     }
+
+    private static String retrieveSanitizedFuncotation(final Funcotation funcotation, final String manualAnnotationSerializedString) {
+        final String initialString = funcotation.serializeToVcfString(manualAnnotationSerializedString);
+        return StringUtils.replaceEach(initialString, new String[]{",", ";", "=", "\t", "|"}, new String[]{"_%2C_", "_%3B_", "_%3D_", "_%09_", "_%7C_"});
```
Comment (jonn-smith) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/vcfOutput/VcfOutputRenderer.java]: Are you sure the HTML escape sequences are what we want to put in here?  Do we need to preserve all the original values for the escape chars or can we replace them all with some other char and move on (e.g. mapping most restricted chars to `_` and another to something else)?

File: src/main/java/org/broadinstitute/hellbender/utils/codecs/gencode/GencodeGtfCodec.java [hunk 5]
```diff
@@ -63,7 +63,7 @@
     static final Logger logger = LogManager.getLogger(GencodeGtfCodec.class);
 
     public static final int GENCODE_GTF_MIN_VERSION_NUM_INCLUSIVE = 19;
-    public static final int GENCODE_GTF_MAX_VERSION_NUM_INCLUSIVE = 27;
+    public static final int GENCODE_GTF_MAX_VERSION_NUM_INCLUSIVE = 28;
```
Comment (jonn-smith) [inline src/main/java/org/broadinstitute/hellbender/utils/codecs/gencode/GencodeGtfCodec.java]: Based on our conversation earlier today, I think it makes sense to remove the `GENCODE_GTF_MAX_VERSION_NUM_INCLUSIVE` field altogether and just attempt to use the specified gencode.  

We can add in a warning / exception when an error occurs to investigate the Gencode version.

File: src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java [hunk 6]
```diff
@@ -29,16 +33,19 @@
 
     // Whether to do debug output (i.e. leave output around).
     // This should always be false when checked in.
-    private static final boolean doDebugTests           = false;
-    private static final String  LARGE_DATASOURCES_FOLDER = "funcotator_dataSources_latest";
+    private static final boolean doDebugTests = false;
+    private static final String LARGE_DATASOURCES_FOLDER = "funcotator_dataSources_latest";
+
+    private static final String PIK3CA_VCF_HG19 = toolsTestDir + "funcotator/0816201804HC0_R01C01.pik3ca.vcf";
```
Comment (jonn-smith) [inline src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java]: I'm not sure this file is publicly releasable - I believe it needs to be sanitized before inclusion in our tests.

File: src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java [hunk 7]
```diff
@@ -336,12 +341,81 @@ public void exhaustiveArgumentTest(final String dataSourcesPath,
...
+
+        runCommandLine(arguments);
+
+        final List<VariantContext> variantContexts = new ArrayList<>();
+        final FeatureDataSource<VariantContext> featureDataSource = new FeatureDataSource<>(outputFile);
+        for (final VariantContext vc : featureDataSource) {
+            variantContexts.add(vc);
+        }
+
+        int NUM_VARIANTS = 21;
```
Comment (jonn-smith) [inline src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java]: final variables

File: src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java [hunk 8]
```diff
@@ -336,12 +341,81 @@ public void exhaustiveArgumentTest(final String dataSourcesPath,
...
+
+        runCommandLine(arguments);
+
+        final List<VariantContext> variantContexts = new ArrayList<>();
+        final FeatureDataSource<VariantContext> featureDataSource = new FeatureDataSource<>(outputFile);
+        for (final VariantContext vc : featureDataSource) {
+            variantContexts.add(vc);
+        }
+
+        int NUM_VARIANTS = 100;
```
Comment (jonn-smith) [inline src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java]: final variables

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/vcfOutput/VcfOutputRenderer.java (hunk 4 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/vcfOutput/VcfOutputRenderer.java]: we don't actually support percent encoding in vcf yet, it's a 4.3 feature and we're only up to 4.2

**Comment by LeeTL1220** [comment]
@jonn-smith @lbergelson Why wouldn't it be publicly releasable?  It's NA12878.  I also confirmed with one of the engineers that there is no issue.

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java (hunk 1 above)
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java (hunk 3 above)
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/vcfOutput/VcfOutputRenderer.java (hunk 4 above)
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/vcfOutput/VcfOutputRenderer.java]: @jonn-smith Previous experience has taught me that the mapping should be invertible.  I'm open to suggestions for changing this...

No action (for now).

File: src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java (hunk 6 above)
Comment (LeeTL1220) [reply src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java]: It is publicly releasable...

No action.

File: src/main/java/org/broadinstitute/hellbender/utils/codecs/gencode/GencodeGtfCodec.java (hunk 5 above)
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/utils/codecs/gencode/GencodeGtfCodec.java]: Done

File: src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java (hunk 7 above)
Comment (LeeTL1220) [reply src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java]: Done

File: src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java (hunk 8 above)
Comment (LeeTL1220) [reply src/test/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorIntegrationTest.java]: Done

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 9]
```diff
@@ -802,6 +802,17 @@ private String createProgramGroupID(final SAMFileHeader header) {
      */
     protected String getToolkitShortName() { return "GATK"; }
 
+    /**
+     * Call {@link GATKTool#addFeatureInputsAfterInitialization(String, String, Class, int)} with no caching.
+     *
+     * @return The {@link FeatureInput} used as the key for this data source.
+     */
+    protected FeatureInput<? extends Feature> addFeatureInputsAfterInitialization(final String filePath, final String name,
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/vcfOutput/VcfOutputRenderer.java (hunk 4 above)
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/vcfOutput/VcfOutputRenderer.java]: I'm assuming that this comment was for our information.  I just wanted a reversible transform. We will probably have to adjust for VCF 4.3.  @lbergelson Should I take any more steps?

No action (for now).

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java [hunk 10]
```diff
@@ -249,9 +248,9 @@
     @Argument(
             fullName = FuncotatorArgumentDefinitions.IGNORE_FILTERED_VARIANTS_LONG_NAME,
             optional = true,
-            doc = "Whether to ignore variants that have been filtered or to annotate them."
+            doc = "Ignore/drop variants that have been filtered in the input."
     )
-    protected boolean includeFilteredVariants = false;
+    protected boolean ignoreFilteredVariants = false;
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java]: :+1: Done

**Comment by LeeTL1220** [comment]
@lbergelson @jonn-smith Regarding references:  I am open to suggestions.  These made life much easier.  We would have to do some work to get rid of them.  Can we file an issue and address later?  Since these are in the large file repo, we do not need to worry about history bloat, correct?

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotation.java [hunk 11]
```diff
@@ -63,6 +63,7 @@ default void setFieldSerializationOverrideValues(final Map<String,String> fieldS
     }
 
     /**
+     * TODO: This should not be specific to a VCF.  That should be the job of the VCFOutputRenderer to sanitize any strings.  A Funcotation should not care whether it is being rendered to a VCF or MAF.
```
[the remark this answers is not in the document]
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotation.java]: I just forgot to file the issue.  Done...
#4797

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java (hunk 2 above)
Comment (LeeTL1220) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/Funcotator.java]: I am only going to file an issue for this.

https://github.com/broadinstitute/gatk/issues/4798

Done.

**Comment by LeeTL1220** [comment]
@jonn-smith @lbergelson Back to you guys.  I have a couple of questions for y'all.

**Comment by LeeTL1220** [comment]
@lbergelson @jonn-smith As per offline discussion, I will targz the references and untar as needed.

**Comment by LeeTL1220** [comment]
@jonn-smith @lbergelson Back to you.

File: src/test/java/org/broadinstitute/hellbender/utils/test/FuncotatorReferenceTestUtils.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    private static String extractFastaTarGzToTemp(final File fastaTarGz, final Path destDir) {
+        Utils.nonNull(fastaTarGz);
+        Utils.nonNull(destDir);
+        String result = null;
+        try (final ArchiveInputStream i = new TarArchiveInputStream(new GzipCompressorInputStream(new FileInputStream(fastaTarGz)))) {
+            ArchiveEntry entry = null;
+            while ((entry = i.getNextEntry()) != null) {
+                if (!i.canReadEntryData(entry)) {
+                    // log something?
```
Comment (jonn-smith) [inline src/test/java/org/broadinstitute/hellbender/utils/test/FuncotatorReferenceTestUtils.java]: Yeah - we should log something in this case.  The tests could be affected by this, but may not be.

Comment (LeeTL1220) [reply src/test/java/org/broadinstitute/hellbender/utils/test/FuncotatorReferenceTestUtils.java]: Done

**Pull request merged**